llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp

```diff
@@ -756,7 +759,8 @@ BitVector SIRegisterInfo::getReservedRegs(const MachineFunction &MF) const {
   }
 
   // Reserve all the AGPRs if there are no instructions to use it.
-  if (!ST.hasMAIInsts())
+  if (!ST.hasMAIInsts() || (ST.hasGFX90AInsts() && !MFI->mayNeedAGPRs() &&
```

Comment:
This should be implicit based on the register count and not need explicit handling